sycl/test/reduction/reduction_transparent.cpp

```diff
@@ -54,12 +66,62 @@ void test(T Identity, size_t WGSize, size_t NWItems) {
   }
 }
 
+// Checks reductions initialized with transparent functor and identity
+// value not explicitly specified. The parameter 'Identity' is passed here
+// only to pre-initialize input data correctly.
+template <typename T, int Dim, class BinaryOperation>
+void testNoId(T Identity, size_t WGSize, size_t NWItems) {
+  buffer<T, 1> InBuf(NWItems);
+  buffer<T, 1> OutBuf(1);
+
+  // Initialize.
+  BinaryOperation BOp;
+  T CorrectOut;
+  initInputData(InBuf, CorrectOut, Identity, BOp, NWItems);
+
+  // Compute.
+  queue Q;
+  Q.submit([&](handler &CGH) {
+    auto In = InBuf.template get_access<access::mode::read>(CGH);
+    accessor<T, Dim, access::mode::discard_write, access::target::global_buffer>
+        Out(OutBuf, CGH);
+
+    range<1> GlobalRange(NWItems);
+    range<1> LocalRange(WGSize);
+    nd_range<1> NDRange(GlobalRange, LocalRange);
+    CGH.parallel_for<SomeNoIdClass<T, Dim, BinaryOperation>>(
+        NDRange, intel::reduction(Out, BOp), [=](nd_item<1> NDIt, auto &Sum) {
+          Sum.combine(In[NDIt.get_global_linear_id()]);
+        });
+  });
+
+  // Check correctness.
+  auto Out = OutBuf.template get_access<access::mode::read>();
+  T ComputedOut = *(Out.get_pointer());
+  if (ComputedOut != CorrectOut) {
+    std::cout << "NWItems = " << NWItems << ", WGSize = " << WGSize << "\n";
+    std::cout << "Computed value: " << ComputedOut
+              << ", Expected value: " << CorrectOut << "\n";
+    assert(0 && "Wrong value.");
+  }
+}
+
+template <typename T, int Dim, class BinaryOperation>
+void test(T Identity, size_t WGSize, size_t NWItems) {
+  testId<T, Dim, BinaryOperation>(Identity, WGSize, NWItems);
+  testNoId<T, Dim, BinaryOperation>(Identity, WGSize, NWItems);
+}
+
 int main() {
 #if __cplusplus >= 201402L
-  test<double, 0, intel::maximum<>>(std::numeric_limits<double>::min(), 7, 7 * 5);
+  test<double, 0, intel::maximum<>>((std::numeric_limits<double>::min)(), 7, 7 * 5);
```

Comment:
Should this be `std::numeric_limits<double>::lowest()` instead?

See https://en.cppreference.com/w/cpp/types/numeric_limits/lowest vs https://en.cppreference.com/w/cpp/types/numeric_limits/min.

If you agree, we should make the same change for the identity for `IsMaximumIdentityOp` here: https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/intel/reduction.hpp#L177

Reply:
That becomes interesting and less trivial...
I agree that Identity set to numeric_limits<double>::lowest is better than numeric_limits<double>::min for MAX-search reductions.
That also made me think about 2 different questions below:

1) What if ALL elements in the input array are -Inf ? If we set the Identity value to numeric_limits::lowest, then after MAX-search the result would be numeric_limits::lowest, which is incorrect.
Possible solution for FP types (and MAX operation) is:
If (std::numeric_limits<T>::has_infinity)
  Identity = -std::numeric_limits<T>::infinity;
else
  Identity = std::numeric_limits<T>::lowest;

2) It seems now that it is quite easy for user to pass identity to reduction constructor which is not equal to what is expected by our implementation, and get a runtime assert printed here: https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/intel/reduction.hpp#L362

I think I am going to remove that assert as it seems to severe now.

Reply:
You're right, I hadn't thought about infinity.  I agree with your proposed solution.

Please do go ahead and remove the assert for now.  We need to think this case through some more -- an alternative way to solve the mismatched identity problem would be to completely disallow specifying an identity if one can be automatically detected (i.e. for the built-in operators and built-in types).  This behavior wouldn't rely on the user knowing what the implementation thinks the identity should be, only that the implementation should know the identity (which is guaranteed by the specification for built-in types and operators).

But before we make that change, we should be sure there aren't any good use-cases for developers overriding the identity value for a standard operator.

Reply:
I fixed this in the 2nd commit in this PR: https://github.com/intel/llvm/pull/1709/commits/7057ebe51dc0d07b28bf4133acd20aedae79bc3a

Reply:
Thanks.

@rolandschulz, @jasonsewall-intel: We should think about whether a specification change would be useful here.

Reply:
This is just my opinion here. I think it is useful to keep the constructor accepting identity value even though the identity value from user may be ignored there. Here is why (copy-paste of the comment from the sources):
```
    // The SPEC could prohibit passing identity parameter to operations with
    // known identity, but that could have some bad consequences too.
    // For example, at some moment the implementation may NOT know the identity
    // for COMPLEX-PLUS reduction. User may create a program that would pass
    // COMPLEX value (0,0) as identity for PLUS reduction. At some later moment
    // when the implementation starts handling COMPLEX-PLUS as known operation
    // the existing user's program remains compilable and working correctly.
    // I.e. with this constructor here, adding more reduction operations to the
    // list of known operations does not break the existing programs.
```

Reply:
I agree that this could be useful, and we should merge what you've implemented.

The reason I think that this may still be a specification issue is that we might need to change the wording to permit this.  There might be weird corner cases where this behavior is observable (e.g. the value written to a `discard_write` accessor may end up being the identity chosen by the implementation instead of the identity supplied by the programmer).

We might need to tighten the wording to say that the identity value provided *may* be used to initialize temporary values, but that an implementation is free to not use it.

Reply:
I prefer if we don't allow identify parameter for known parameter. If you do you either need to use the one provided by the user, which has a performance impact. Or you don't but now you have to add an assert but that only helps in debug mode. I think the problem with the updated API can be better addressed with automatic refactoring (e.g. clang-tidy). They are easy to write for simple cases like this.

I agree that we should be using `-infinity` not `minimum`.